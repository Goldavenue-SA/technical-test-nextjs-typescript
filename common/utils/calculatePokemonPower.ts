import { Pokemon } from '../interfaces/pokemon';

export const calculatePokemonPower = (pokemon: Pokemon) =>
  pokemon.hp +
  pokemon.speed +
  pokemon.attack +
  pokemon.special_attack +
  pokemon.defense +
  pokemon.special_defense;
