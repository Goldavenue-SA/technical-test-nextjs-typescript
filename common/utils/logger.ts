import winston from 'winston';

const createLogger = () => {
  if (typeof window !== 'undefined') {
    return winston.createLogger({
      level: 'info',
      format: winston.format.combine(
        winston.format.timestamp(),
        winston.format.json()
      ),
      transports: [
        new winston.transports.Console({
          format: winston.format.simple(),
        }),
      ],
    });
  }

  const logger = winston.createLogger({
    level: 'info',
    format: winston.format.combine(
      winston.format.timestamp(),
      winston.format.json()
    ),
    transports: [
      new winston.transports.Console({
        format: winston.format.simple(),
      }),
    ],
  });

  if (process.env.NODE_ENV === 'development') {
    logger.add(
      new winston.transports.File({
        filename: 'logs/error.log',
        level: 'error',
      })
    );
    logger.add(
      new winston.transports.File({
        filename: 'logs/combined.log',
      })
    );
  }

  return logger;
};

export const logger = createLogger();
