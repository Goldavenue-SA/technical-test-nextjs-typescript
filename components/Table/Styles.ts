import styled from 'styled-components';

const StyledTable = styled.table`
  width: 100%;
  max-width: 900px;
  margin: 20px auto;
  border-spacing: 0;
  border-radius: 12px;
  background: rgba(255, 255, 255, 0.08);
  box-shadow: 0 4px 10px rgba(0, 0, 0, 0.1);
  backdrop-filter: blur(10px);
  overflow: hidden;
`;

const TableHeader = styled.thead`
  background: rgba(255, 255, 255, 0.15);

  th {
    padding: calc(var(--global-rhythm) / 1.5);
    font-weight: 600;
    font-size: 1rem;
    text-align: left;
  }

  tr:first-child th:first-child {
    border-top-left-radius: 12px;
  }
  tr:first-child th:last-child {
    border-top-right-radius: 12px;
  }
`;

const TableBody = styled.tbody`
  tr {
    border-bottom: 1px rgba(255, 255, 255, 1);

    &:hover {
      background: rgba(255, 255, 255, 0.12);
      cursor: pointer;
    }
  }

  td {
    padding: calc(var(--global-rhythm) / 1.5);
    font-size: 0.95rem;
    text-align: left;
    background: rgba(255, 255, 255, 0.05);
  }
`;

const TableFooter = styled.tfoot`
  td {
    padding: 12px;
    font-weight: 500;
    text-align: right;
    background: rgba(255, 255, 255, 0.1);

    &:first-child {
      border-bottom-left-radius: 12px;
    }
    &:last-child {
      border-bottom-right-radius: 12px;
    }
  }
`;

export { StyledTable, TableBody, TableFooter, TableHeader };
