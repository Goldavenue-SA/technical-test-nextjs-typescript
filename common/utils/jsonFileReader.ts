import path from 'path';
import fs from 'fs';
import { logger } from './logger';

export const readJsonFile = async <T>(filePath: string): Promise<T> => {
  try {
    const jsonFilePath = path.join(process.cwd(), filePath);
    const jsonData = await fs.promises.readFile(jsonFilePath, 'utf-8');

    return JSON.parse(jsonData);
  } catch (error) {
    logger.error('Error reading JSON file:', error);

    throw new Error('Failed to read or parse the JSON file.');
  }
};
