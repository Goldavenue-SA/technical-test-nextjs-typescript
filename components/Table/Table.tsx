import { Pokemon } from '../../common/interfaces/pokemon';
import { calculatePokemonPower } from '../../common/utils/calculatePokemonPower';
import { StyledTable, TableBody, TableFooter, TableHeader } from './Styles';

export interface TableProps {
  pokemonList: Pokemon[];
}

export const Table = ({ pokemonList = [] }: TableProps) => {
  return (
    <>
      <StyledTable>
        <TableHeader>
          <tr>
            <th>ID</th>
            <th>Name</th>
            <th>Type</th>
            <th colSpan={6}>Stats</th>
            <th>Power</th>
          </tr>
        </TableHeader>
        <TableBody>
          {pokemonList.map((pokemon, index) => (
            <tr key={`${pokemon.name}${index}`}>
              <td>{pokemon.id}</td>
              <td>{pokemon.name}</td>
              <td>{pokemon.type.join(', ')}</td>
              <td>{pokemon.hp}</td>
              <td>{pokemon.speed}</td>
              <td>{pokemon.attack}</td>
              <td>{pokemon.special_attack}</td>
              <td>{pokemon.defense}</td>
              <td>{pokemon.special_defense}</td>
              <td>{calculatePokemonPower(pokemon)}</td>
            </tr>
          ))}
        </TableBody>
      </StyledTable>
    </>
  );
};
