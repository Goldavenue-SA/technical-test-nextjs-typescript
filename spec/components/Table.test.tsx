jest.mock('../../common/utils/calculatePokemonPower', () => ({
  calculatePokemonPower: jest.fn().mockReturnValue(100),
}));

import { render, screen } from '@testing-library/react';
import { Table, TableProps } from '../../components/Table/Table';
import { calculatePokemonPower } from '../../common/utils/calculatePokemonPower';
import '../setupTests';

describe('Test Table Component', () => {
  const tableProps: TableProps = {
    pokemonList: [
      {
        id: 1,
        name: 'Bulbasaur',
        type: ['Grass', 'Water'],
        hp: 45,
        attack: 49,
        defense: 48,
        special_attack: 65,
        special_defense: 67,
        speed: 46,
      },
      {
        id: 2,
        name: 'Ivysaur',
        type: ['Poison', 'Fire'],
        hp: 60,
        attack: 62,
        defense: 63,
        special_attack: 81,
        special_defense: 80,
        speed: 69,
      },
    ],
  };

  const setupComponent = () => render(<Table {...tableProps} />);

  it('Renders correctly and matches the snapshot', () => {
    const { asFragment } = setupComponent();

    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render the component with the correct props', () => {
    setupComponent();

    expect(screen.getByText('ID')).toBeInTheDocument();
    expect(screen.getByText('Name')).toBeInTheDocument();
    expect(screen.getByText('Type')).toBeInTheDocument();
    expect(screen.getByText('Stats')).toBeInTheDocument();
    expect(screen.getByText('Power')).toBeInTheDocument();
  });

  it('Renders data rows correctly for each Pokemon', () => {
    setupComponent();

    const rows = screen.getAllByRole('row');

    expect(rows[1]).toHaveTextContent('1');
    expect(rows[1]).toHaveTextContent('Bulbasaur');
    expect(rows[1]).toHaveTextContent('Grass, Water');
    expect(rows[1]).toHaveTextContent('45');
    expect(rows[1]).toHaveTextContent('49');
    expect(rows[1]).toHaveTextContent('48');
    expect(rows[1]).toHaveTextContent('65');
    expect(rows[1]).toHaveTextContent('67');
    expect(rows[1]).toHaveTextContent('46');
    expect(rows[1]).toHaveTextContent('100');

    expect(rows[2]).toHaveTextContent('2');
    expect(rows[2]).toHaveTextContent('Ivysaur');
    expect(rows[2]).toHaveTextContent('Poison, Fire');
    expect(rows[2]).toHaveTextContent('60');
    expect(rows[2]).toHaveTextContent('62');
    expect(rows[2]).toHaveTextContent('63');
    expect(rows[2]).toHaveTextContent('81');
    expect(rows[2]).toHaveTextContent('80');
    expect(rows[2]).toHaveTextContent('69');
    expect(rows[2]).toHaveTextContent('100');
  });

  it('Calls calculatePokemonPower for each PokÃ©mon', () => {
    setupComponent();

    expect(calculatePokemonPower).toHaveBeenCalledWith(
      tableProps.pokemonList[0]
    );
    expect(calculatePokemonPower).toHaveBeenCalledWith(
      tableProps.pokemonList[1]
    );
  });

  test('Renders no rows when pokemonList is empty', () => {
    render(<Table pokemonList={[]} />);

    const rows = screen.queryAllByRole('row');

    expect(rows).toHaveLength(1);
  });
});
