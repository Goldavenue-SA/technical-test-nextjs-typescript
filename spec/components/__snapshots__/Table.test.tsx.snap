// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Table Component Renders correctly and matches the snapshot 1`] = `
<DocumentFragment>
  <table
    class="Styles__StyledTable-sc-1o1cw3u-0 UzSlY"
  >
    <thead
      class="Styles__TableHeader-sc-1o1cw3u-1 jSvCgs"
    >
      <tr>
        <th>
          ID
        </th>
        <th>
          Name
        </th>
        <th>
          Type
        </th>
        <th
          colspan="6"
        >
          Stats
        </th>
        <th>
          Power
        </th>
      </tr>
    </thead>
    <tbody
      class="Styles__TableBody-sc-1o1cw3u-2 jMVdeV"
    >
      <tr>
        <td>
          1
        </td>
        <td>
          Bulbasaur
        </td>
        <td>
          Grass, Water
        </td>
        <td>
          45
        </td>
        <td>
          46
        </td>
        <td>
          49
        </td>
        <td>
          65
        </td>
        <td>
          48
        </td>
        <td>
          67
        </td>
        <td>
          100
        </td>
      </tr>
      <tr>
        <td>
          2
        </td>
        <td>
          Ivysaur
        </td>
        <td>
          Poison, Fire
        </td>
        <td>
          60
        </td>
        <td>
          69
        </td>
        <td>
          62
        </td>
        <td>
          81
        </td>
        <td>
          63
        </td>
        <td>
          80
        </td>
        <td>
          100
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
